Implement PEP561

ethanhs opened the pull request:
This is an implementation of PEP 561. I am picking up from https://github.com/python/mypy/pull/4403 because I'd rather not rebase over 100 commits (and some of the changes shouldn't go in!). Third times a charm I hope :)

Tasks:
- [x] Test PEP 561 implementation
- [x] Check PEP 561 conformance to [resolution order](https://www.python.org/dev/peps/pep-0561/#type-checker-module-resolution-order)
- [x] support running with alternate Python executable ``--python-executable``
- [x] document PEP 561 features
- [x] document ``--python-executable`` flag
- [x] test ``--python-executable`` flag
- [x] test ``--no-site-packages`` flag
- [x] document `--no-site-packages` flag
- [x] `--python-version` flag sets `--python-executable` and visa-versa if possible

This branch likely works as intended and is feature complete, but it is possible that I've overlooked something/made a mistake.

Fixes #2625, #1190, #965.

eric-wieser commented on mypy/build.py:
```diff
@@ -804,7 +806,7 @@ def remove_cwd_prefix_from_path(p: str) -> str:
 
 
 # Cache find_module: (id, lib_path) -> result.
```
This comment needs updating

eric-wieser commented on mypy/build.py:
```diff
@@ -868,59 +870,133 @@ def is_file(path: str) -> bool:
     """Return the path of the module source file, or None if not found."""
     lib_path = tuple(lib_path_arg)
-
-    def find() -> Optional[str]:
```
Can you make this refactor in a separate PR too?

ethanhs commented:
I just reverted some of the more significant refactoring started in the last PR so that this is easier to review. Most of it will be refactored later anyway.

eric-wieser commented on mypy/main.py:
```diff
@@ -205,6 +207,72 @@ def invert_flag_name(flag: str) -> str:
+        options.python_version = _python_version_from_executable(
+            special_opts.python_executable)
+        options.python_executable = special_opts.python_executable
+    return options
```
Perhaps misleading that this modifies in place _and_ returns the modified value

ethanhs replied:
Are you suggesting that the comment at the beginning might be misleading? (That could be, I'm just not sure what you mean by this comment).

eric-wieser replied:
I'd expect either
```
options.python_version, options.python_executable = infer_python_version_and_executable(special_opts)
```
or
```
infer_python_version_and_executable(options, special_opts)
```

But not
```
options = infer_python_version_and_executable(options, special_opts)
```

(think how `list.sort()`, `dict.update()`, etc, all modify in place and do not return the new result)

eric-wieser commented on mypy/build.py:
```diff
@@ -891,36 +930,65 @@ def find() -> Optional[str]:
+                path = os.path.join(pkg_dir, dir_chain)
+                third_party_dirs.append(path)
+
+        candidate_base_dirs = tuple(third_party_dirs + find_module_dir_cache[dir_chain, lib_path])
```
Why the tuple conversion?

ethanhs replied:
Ah, this is just due to it being in my previous commits. I will get rid of it when I probably end up rebasing this on #4623.

eric-wieser replied:
This looks pretty close to done to me - looks like your patch got a lot cleaner after adding the refactor and removing it again :). So maybe it will get in before #4623 after all?

ethanhs replied:
Not sure. It is probably easier for me to rebase since my implementation is just one commit to rollback and apply, but I would prefer to get this PR in first. :)

eric-wieser commented on mypy/build.py:
```diff
@@ -891,36 +930,65 @@ def find() -> Optional[str]:
+
+        if python_executable is not None:
+            site_packages_dirs = get_site_packages_dirs(python_executable)
+            if not site_packages_dirs:
```
Should this be `site_packages_dirs is not None`?

ethanhs replied:
No, I just realized it is dead code.

eric-wieser commented on mypy/test/testargs.py:
```diff
@@ -17,3 +22,47 @@ def test_coherence(self) -> None:
+
+    def test_executable_inference(self) -> None:
+        """Test the --python-executable flag with --python-version"""
+        sys_ver_str = '.'.join(map(str, sys.version_info[:2]))
```
nit: possibly more readable as `'{v.major}.{v.minor}'.format(sys.version_info)`

eric-wieser commented on mypy/test/testargs.py:
```diff
@@ -17,3 +22,47 @@ def test_coherence(self) -> None:
+        with pytest.raises(PythonExecutableInferenceError) as e:
+            options = infer_python_version_and_executable(options, special_opts)
+        assert str(e.value) == 'Python version (2, 10) did not match executable {}, got' \
+                               ' version {}.'.format(sys.executable, str(sys.version_info[:2]))
```
nit: No need for the `str` here

ethanhs replied:
It is indeed needed. `type(e.value)` is `mypy.main.PythonExecutableInferenceError`.

eric-wieser replied:
I meant the second `str` on `sys.version_info`

eric-wieser commented on mypy/build.py:
```diff
@@ -868,7 +870,46 @@ def is_file(path: str) -> bool:
+            # the python executable is likely in a virtual environment, thus lacking
+            # needed site methods
+            output = call_python(python_executable, VIRTUALENV_SITE_PACKAGES)
+    return [line for line in output.splitlines() if os.path.isdir(line)]
```
Why do you check the directories exist in this path, but not in the `python_executable == sys.executable` path?

eric-wieser commented on mypy/build.py:
```diff
@@ -891,36 +932,57 @@ def find() -> Optional[str]:
     if key not in find_module_cache:
+        components = id.split('.')
+        dir_chain = os.sep.join(components[:-1])  # e.g., 'foo/bar'
+        site_packages_dirs = get_site_packages_dirs(python_executable)
```
Is there any reason not to put these three lines inside `find`, which makes the flow a little clearer?

ethanhs replied:
Not really. They originally were broken out due to the refactor, but I agree putting them in find is ideal right now.

eric-wieser commented on mypy/build.py:
```diff
@@ -877,10 +918,10 @@ def find() -> Optional[str]:
-            dirs = []
+        # TODO (ethanhs): refactor to use lru_cache on each of these searches
+        dirs = find_module_dir_cache.get((dir_chain, lib_path), [])
+        if not dirs:
```
This is not equivalent to the previous code. Your new code ignores the cache if if says that `dirs == []`, which seems undesirable.

eric-wieser commented on mypy/build.py:
```diff
@@ -891,36 +932,57 @@ def find() -> Optional[str]:
         # elements of lib_path.  This is probably much shorter than lib_path itself.
         # Now just look for 'baz.pyi', 'baz/__init__.py', etc., inside those directories.
-        seplast = os.sep + components[-1]  # so e.g. '/baz'
+        seplast = os.sep + components[-1]
```
Was this comment inaccurate?

eric-wieser commented on mypy/build.py:
```diff
@@ -868,7 +870,46 @@ def is_file(path: str) -> bool:
-def find_module(id: str, lib_path_arg: Iterable[str]) -> Optional[str]:
+USER_SITE_PACKAGES = \
+    'from __future__ import print_function; import site; print(site.getusersitepackages());' \
+    'print(*site.getsitepackages(), sep="\\n")'
```
This fails in the corner-case where the paths contain newlines(!)

Better would be to do

```
print(repr(site.getsitepackages() + [site.getusersitepackages()]))
```
and then read it with `ast.literal_eval`

eric-wieser replied:
I also worry that you might have a different between the stdout encoding and the default `.decode()` decoding, which using `repr`/`ast.literal_eval` or maybe even `pickle` would resolve.

Are site package paths unicode, bytes, or either?

ethanhs replied:
Should be `str`, for each meaning in 2 and 3, so your recommendation is well taken.

gvanrossum commented:
I am going to snooze this PR until the chatter dies down and it has been
rebased on top of #4623 (which isn't quite ready to land yet, but will be
later this week).

ethanhs commented:
Okay now that #4623 is merged I will revert the last few commits I have made and hand-reapply them.

gvanrossum commented:
SGTM.

ethanhs commented:
The travis errors are bizarre. I restarted the 3.7 job to test if they are flakes.

ethanhs commented:
Hm, this is really strange it seems it is https://github.com/pytest-dev/pytest-xdist/issues/81 ? But I don't think I exit early anywhere, so I'm not sure what is going on.

eric-wieser commented on mypy/build.py:
```diff
@@ -877,26 +935,31 @@ def _find_module(self, id: str, lib_path: Tuple[str, ...]) -> Optional[str]:
         lib_path = tuple(lib_path_arg)
 
-        key = (id, lib_path)
+        key = (id, python_executable, lib_path)
```
Nit: make the order match the argument order.

ethanhs commented:
Also it works fine on Windows, so maybe this is Travis and/or Linux specific?

ethanhs commented:
I checked on Linux and the tests pass locally there as well. The failures seem to be some quirk of Travis. I'm going to try restarting them through a push.

ethanhs commented:
Hm, it still is failing. Really weird. I'm just about stumped right now.

eric-wieser commented:
Try removing the `-j12` from travis to remove the parallelism?

ethanhs commented:
Yeah I will try that now.

ethanhs commented:
Interestingly, on the latest run it is reporting `OSError: [Errno 12] Cannot allocate memory`. I have no idea why it would OOM.

```
_____________________________ testReadOnlyProperty _____________________________
[gw0] linux -- Python 3.4.6 /home/travis/virtualenv/python3.4.6/bin/python
data: /home/travis/build/python/mypy/test-data/unit/pythoneval.test:323:
/home/travis/build/python/mypy/mypy/test/testpythoneval.py:42: in run_case
    test_python_evaluation(testcase)
/home/travis/build/python/mypy/mypy/test/testpythoneval.py:84: in test_python_evaluation
    returncode, interp_out = run_command([interpreter, program])
/home/travis/build/python/mypy/mypy/test/helpers.py:368: in run_command
    cwd=cwd,
/opt/python/3.4.6/lib/python3.4/subprocess.py:856: in __init__
    restore_signals, start_new_session)
/opt/python/3.4.6/lib/python3.4/subprocess.py:1398: in _execute_child
    restore_signals, start_new_session, preexec_fn)
E   OSError: [Errno 12] Cannot allocate memory
_________________________ mypy/test/testpythoneval.py __________________________
```

eric-wieser commented:
Caches getting obscenely large?

ethanhs commented:
>Caches getting obscenely large?

I mean I suppose that is the only reasonable explanation, but I'm not sure how that could be. I'll have to compare this branch to past implementations to see if there are any important differences.

eric-wieser commented:
It could be a bug in the commits that you rebased upon

eric-wieser commented:
Perhaps you could inspect the size of the dicts in FileSystemMetaCache?

ethanhs commented:
As an update, I've been busy with midterms recently, but should be able to get back to this tomorrow.

ethanhs commented:
So I was looking at memory usage, and it appears that it jumps to about 750MB before even running the first test! At some point it may be worth it to look at what is causing this, as this ends up being about 40% of current memory usage in the test suite (we end up at ~1.8GB).

In this branch the memory usage hits about 4.1GB. I have yet to find why this is the case. The results cache seems to not contribute to the constant memory increase that occurs as the tests continue.

gvanrossum commented:
What did you use to measure those numbers?

ethanhs commented:
I was using the built in Windows Task Manager. I also began look at each cache's size, but the `results` and `dirs` caches seem not to be the biggest impact. I am going to look at how large the FS caches get now.

ethanhs commented:
Okay, I tried moving things over to using lru_cache in most every place, and making the cache size zero (disabling both fscache and the find module caches). Now I can start moving the maxsize for each cache up until I am happy. This doesn't seem like an ideal solution, but it seems to be the only one I can find.

ethanhs commented on mypy/fscache.py:
```diff
@@ -46,31 +47,13 @@ def flush(self) -> None:
         self.listdir_error_cache = {}  # type: Dict[str, Exception]
```
I have a local commit that replaces these to use cache_clear on the `lru_cache` decorated methods, but I didn't want to waste CI time by pushing it yet. I will do so when I have some good results from increasing the cache sizes.

eric-wieser commented:
Note that using `lru_cache` on methods [might be a bad idea](https://stackoverflow.com/q/33672412/102441)

ethanhs commented:
>Note that using lru_cache on methods might be a bad idea

@eric-wieser duly noted, I will take a look at using a wrapper around `lru_cache` to avoid this issue.

So, for some reason the changes I have made "fixed" the memory issues (though I am not entirely sure why?), without needing to limit cache sizes. 

One change I'm pretty sure I want to make is remove caching from `fscache.isfile_case` as that method almost always misses.

ethanhs commented:
@eric-wieser I took a look at the issue with `lru_cache` on methods, and I'm not convinced that it is big enough to write and maintian a wraper around `lru_cache` for either of these cases. The caches for the file system and `find_module` both will stay alive essentially for the life of the program, so the lifetime of the caches/classes isn't an issue IMO.

eric-wieser commented:
The other issue is that `.flush()` clears the cache for _all_ instances, not just for the current one.

Marking it as a `@classmethod` would make that at least not be surprising.

eric-wieser commented:
A trivial fix would be

```
def __init__(self):
    self.stat = lru_cache(None)(self.stat)
    self.list_dir = lru_cache(None)(self.list_dir)
```

eric-wieser commented on mypy/fscache.py:
```diff
@@ -40,37 +41,16 @@ def __init__(self) -> None:
-        self.stat_cache = {}  # type: Dict[str, os.stat_result]
-        self.stat_error_cache = {}  # type: Dict[str, Exception]
-        self.listdir_cache = {}  # type: Dict[str, List[str]]
-        self.listdir_error_cache = {}  # type: Dict[str, Exception]
```
Note that [`lru_cache` does not support this feature](https://bugs.python.org/issue21537)

ethanhs replied:
Yes I will be adding an exception cache next. I'm not entirely sure caching exceptions is a good idea (since this could lead to difficulties debugging), but I should probably leave as is.

JukkaL replied:
Caching exceptions is needed for transactional semantics. If we don't cache them, the second call targeting the same path could succeed. (Also, performance could be impacted if we perform the same failing operation many times, but this is perhaps unlikely.)

ethanhs commented:
This should be ready for another round of review (thanks!).

gvanrossum commented:
Thanks, @JukkaL will take over this review.

JukkaL commented on docs/source/installed_packages.rst:
```diff
@@ -0,0 +1,114 @@
+        packages=["package_c-stubs"]
+    )
+
+Using PEP 561 compatible packages with mypy
```
I'd prefer to have this section the first section since likely many more users will be consuming PEP 561 packages than will be creating them.

JukkaL commented on docs/source/installed_packages.rst:
```diff
@@ -0,0 +1,114 @@
+========================
+
+`PEP 561 <https://www.python.org/dev/peps/pep-0561/>`_ specifies how to mark
+a package as supporting type checking. Below is a summary of how to create
```
I'd suggest adding some more background information here for those users that have minimal idea of how packaging in Python works. The goal would be that most users won't need to actually read PEP 561. Some ideas (maybe add a new section with background information or include these in the introduction):

* Write something like "a package *installed using pip, for example* ..." to make it more explicit what sort of packages we are talking about (the term 'package' is an overloaded term).
* Explain what "supporting type checking" means and why it's useful. This can mean that the package has inline type annotations that mypy can use to type check code that imports the package, or that the package ships stub files that mypy can use.
* Also contrast this to typeshed stubs and discuss which stubs take precedence in case stubs exist both in typeshed and in an installed package.
* Add note about the possibility of stubs being shipped in a separate stubs package (which might not be owned by the author of the original package).

JukkaL commented on docs/source/installed_packages.rst:
```diff
@@ -0,0 +1,114 @@
+Making PEP 561 compatible packages
+**********************************
+
+Packages that must be imported at runtime and supply type information should
```
It's not entirely clear what "packages that must be imported at runtime" means. Can you rephrase it? Does this means packages with type annotations included directly in the package implementation?

JukkaL commented on docs/source/installed_packages.rst:
```diff
@@ -0,0 +1,114 @@
+    )
+
+In this example, both ``lib.py`` and ``lib.pyi`` exist. At runtime, ``lib.py``
+will be used, however mypy will use ``lib.pyi``.
```
Style nit: avoid the use of comma before 'however'. Maybe write something like this: "At runtime, the Python interpreter will use lib.py, but mypy will use lib.pyi instead."

JukkaL commented on mypy/build.py:
```diff
@@ -809,6 +813,44 @@ def remove_cwd_prefix_from_path(fscache: FileSystemCache, p: str) -> str:
+        # executable
+        try:
+            output = call_python(python_executable, USER_SITE_PACKAGES)
+        except subprocess.CalledProcessError:
```
Maybe we could have the try/expect in the executed Python fragment to avoid the performance penalty of starting a Python subprocess twice?

JukkaL commented on mypy/build.py:
```diff
@@ -809,6 +813,44 @@ def remove_cwd_prefix_from_path(fscache: FileSystemCache, p: str) -> str:
+    'from __future__ import print_function; import site; print(repr(site.getsitepackages()' \
+    ' + [site.getusersitepackages()]))'
+VIRTUALENV_SITE_PACKAGES = \
+    'from distutils.sysconfig import get_python_lib; print(repr([get_python_lib()]))'
```
Should we do the `__future__` import here as well for consistency, or not do it all?

JukkaL commented on mypy/fscache.py:
```diff
@@ -28,49 +28,45 @@
 class FileSystemMetaCache:
     def __init__(self) -> None:
-        self.flush()
+        self.stat = functools.lru_cache(maxsize=None)(self._stat)
```
It's not clear to me why we are switching to `lru_cache` here -- I haven't followed all the previous discussion.

ethanhs replied:
I did it when I was trying to remove the memory leak. I can try switching back if you think that is better.

ethanhs replied:
I too was curious about traceback usage, so on that pull request [I tested the memory usage](https://github.com/python/mypy/pull/4809#issuecomment-376747428), and found no difference. I suppose the main argument for `lru_cache` is that it requires a bit less boilerplate, but I don't think that is super important. I will try replacing the `lru_cache`s with `dict`s.

JukkaL replied:
What's the current status of the memory leak investigation? Do you believe it may still be happening?

JukkaL commented:
Also, typeshed is causing a merge conflict.

eric-wieser commented on mypy/build.py:
```diff
@@ -809,6 +813,43 @@ def remove_cwd_prefix_from_path(fscache: FileSystemCache, p: str) -> str:
+    print(repr(site.getsitepackages() + [user_dir]))
+else:
+    print(repr([get_python_lib()]))
+'''
```
How about actually reusing the code path?
```python
code = '''
from __future__ import print_function

def main():
	from distutils.sysconfig import get_python_lib
	import site
	if hasattr(site, 'getusersitepackages') and hasattr(site, 'getsitepackages'):
	    user_dir = site.getusersitepackages()
	    return site.getsitepackages() + [user_dir]
	else:
	    return [get_python_lib()]

if __name__ == '__main__':
	print(repr(main()))
'''

if active_python:
    exec_globals = {}
    exec(code, exec_globals)
    res = exec_globals['main']()
else:
    res = ast.literal_eval(call(['python', '-c', code]))
```

ethanhs replied:
That seems rather hacky and slower for little reason.

eric-wieser replied:
A slightly less hacky/slow version would be to put this in it's own file, and do:
```
from . import the_file
if this_python:
    res = the_file.main()
else:
    res = ast.literal_eval(call(['python', '-c', the_file.__file__]))
```

with the rationale of removing code duplication.

ethanhs replied:
I'm not convinced forcing them to be the same makes sense. One is printing the repr of objects, the other is returning the objects. I don't like the idea of coercing code to fit to one or the other just for the sake of symmetry.

eric-wieser replied:
Both are trying to construct exactly the same objects, all that changes is how they return them. One can just return it, the other goes via `repr`/`literal_eval`

eric-wieser replied:
Bear in mind that having `this_python` as a code path at all is just an optimization of the `else` case, so it definitely ought to be symmetric.

ethanhs replied:
I suppose I find the second acceptable. Why would writing bytecode be dangerous? It is versioned for a reason.

eric-wieser commented on mypy/sitepkgs.py:
```diff
@@ -0,0 +1,20 @@
+
+from distutils.sysconfig import get_python_lib
+import site
+from typing import List
```
Should you assume that the python 2 installation has `typing`?

ethanhs replied:
If you are type checking Python 2 code I hope you would have typing installed. Just to be safe I can do:

```
MYPY = False
if MYPY:
    from typing import List
```

JukkaL commented on mypy/main.py:
```diff
@@ -263,10 +329,16 @@ def add_invertible_flag(flag: str,
+                        dest='special-opts:python_executable')
+    parser.add_argument('--no-site-packages', action='store_true',
+                        dest='special-opts:no_executable',
+                        help="Do not search for installed PEP 561 compliant packages.")
```
Remove dot from the end of the help text.

JukkaL commented on mypy/main.py:
```diff
@@ -263,10 +329,16 @@ def add_invertible_flag(flag: str,
     parser.add_argument('--platform', action='store', metavar='PLATFORM',
                         help="typecheck special-cased code for the given OS platform "
-                        "(defaults to sys.platform).")
+                             "(defaults to sys.platform).")
```
Remove dot from the end of the help text.

eric-wieser commented on mypy/build.py:
```diff
@@ -822,20 +842,30 @@ class FindModuleCache:
         self.fscache = fscache or FileSystemMetaCache()
-        # Cache find_module: (id, lib_path) -> result.
-        self.results = {}  # type: Dict[Tuple[str, Tuple[str, ...]], Optional[str]]
+        self.find_lib_path_dirs = functools.lru_cache(maxsize=None)(self._find_lib_path_dirs)
```
I'd argue that the cache-correctness of `lru_cache` is more valuable than the type-checking provided by mypy

eric-wieser commented on mypy/main.py:
```diff
@@ -213,6 +215,70 @@ def invert_flag_name(flag: str) -> str:
+
+if sys.platform == 'win32':
+    def python_executable_prefix(v: str) -> List[str]:
+        return ['py', '-{}'.format(v)]
```
Maybe refer to https://docs.python.org/3/using/windows.html#python-launcher-for-windows in a comment

ethanhs commented on mypy/main.py:
```diff
@@ -213,6 +215,70 @@ def invert_flag_name(flag: str) -> str:
+
+def _python_version_from_executable(python_executable: str) -> Tuple[int, int]:
+    try:
+        check = subprocess.check_output([python_executable, '-c',
```
We only will at most run it twice. (3) is always run, but only one of (1) or (2) is needed to be run. Though perhaps combining them is for the best.

ethanhs commented:
@JukkaL I've made all of the changes you requested save changing `lru_cache`s to `dict`s, since there seems to be some disagreement over which to use. I also wanted to hear more of your thoughts on combining subprocess calls.

JukkaL commented on mypy/build.py:
```diff
@@ -2129,8 +2167,6 @@ def dispatch(sources: List[BuildSource], manager: BuildManager) -> Graph:
                                       for g in graph.values()),
                       graph_load_time=(t1 - t0),
-                      fm_cache_size=len(manager.find_module_cache.results),
-                      fm_dir_cache_size=len(manager.find_module_cache.dirs),
```
If you refactor `lru_cache` uses away, these could be restored. Or maybe these stats could be retrieved from the lru caches?

ethanhs replied:
Yes I should add these back based on the `lru_cache.cache_info()` `currsize`.

ethanhs commented on mypy/test/testpep561.py:
```diff
@@ -0,0 +1,115 @@
+                       expected_returncode: int = 1) -> None:
+        """Helper to run mypy and check the output."""
+        out, err, returncode = mypy.api.run(cmd_line)
+        assert out == expected_out, err
```
Yes, this is so the test output will show both if the assert fails. For example if something breaks, mypy would print `mypy can't find module blah` to stderr, with an empty stdout.

ethanhs commented on mypy/fscache.py:
```diff
@@ -88,8 +84,6 @@ def isfile_case(self, path: str) -> bool:
         TODO: We should maybe check the case for some directory components also,
         to avoid permitting wrongly-cased *packages*.
         """
-        if path in self.isfile_case_cache:
```
It had an extremely high cache miss rate, I probably should have said something. In the tests it seems to miss 99% of the time, so I don't see caching helping.

JukkaL replied:
Tests may not be indicative of real-world performance. More importantly, the cache is not only used for improving performance -- the goal is to provide more transactional semantics for file operations, and omitting caching for some file operations breaks that.

ethanhs commented:
>If you are short on time now, some of the suggestions can be left out and implemented in another PR so that we can merge this sooner.

I left the `lru_cache`s, as that is still somewhat a point of contention (also the memory usage issues disappeared when I put them in, so I'm a little hesitant to go back). I also left the current Python version/executable inference in as-is, which should be addressed at a later date once there is a better understanding of their performance impact.

>If you are able to update the PR by Monday we may just have enough time to include this in the next release

Great! Now that I have updated it, and it will be merged by Monday, perhaps @dmoisset and or @carljm would be interested in testing this? (I realize its a bad time, as it is already kinda late on a Friday...)

>Thanks for persevering with this -- this is going to make a big difference!

Happy to do it, and I want to thank you, Eric, and Jelle for the reviews! They were quite helpful.

dmoisset commented:
I'll try to take a look at this tomorrow... I don't have much time but I'm
very eager to see this merged

On 7 April 2018 at 00:57, Ethan Smith <notifications@github.com> wrote:

> If you are short on time now, some of the suggestions can be left out and
> implemented in another PR so that we can merge this sooner.
>
> I left the lru_caches, as that is still somewhat a point of contention
> (also the memory usage issues disappeared when I put them in, so I'm a
> little hesitant to go back). I also left the current Python
> version/executable inference in as-is, which should be addressed at a later
> date once there is a better understanding of their performance impact.
>
> If you are able to update the PR by Monday we may just have enough time to
> include this in the next release
>
> Great! Now that I have updated it, and it will be merged by Monday,
> perhaps @dmoisset <https://github.com/dmoisset> and or @carljm
> <https://github.com/carljm> would be interested in testing this? (I
> realize its a bad time, as it is already kinda late on a Friday...)
>
> Thanks for persevering with this -- this is going to make a big difference!
>
> Happy to do it, and I want to thank you, Eric, and Jelle for the reviews!
> They were quite helpful.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/mypy/pull/4693#issuecomment-379414275>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAMQHV_Z417Q4e-xVBFy_-CitGAhtfXwks5tmADcgaJpZM4SgTDj>
> .
>



-- 
Daniel F. Moisset - UK Country Manager - Machinalis Limited
www.machinalis.co.uk <http://www.machinalis.com>
Skype: @dmoisset T: + 44 7398 827139

1 Fore St, London, EC2Y 9DT

Machinalis Limited is a company registered in England and Wales. Registered
number: 10574987.

ethanhs commented:
@dmoisset FWIW the release itself is Friday, so you do have some time.
